File: Python/Longest Subarray of 1's After Deleting One Element.py
from typing import List


class Solution:
    """
    Time:   O(n)
    Memory: O(1)
    """

    def longestSubarray(self, nums: List[int]) -> int:
        longest = prev = curr = 0

        for bit in nums:
            if bit:
                curr += 1
            else:
                prev, curr = curr, 0

            longest = max(longest, prev + curr)

        return longest - (longest == len(nums))

File: Java/Longest Subarray of 1's After Deleting One Element.java
class Solution {

    /**
     * Time:   O(n)
     * Memory: O(1)
     */
    public int longestSubarray(int[] nums) {
        int longest = 0;
        int prev = 0, curr = 0;

        for (int bit : nums) {
            if (bit == 0) {
                prev = curr;
                curr = 0;
            } else {
                longest = Math.max(longest, prev + ++curr);
            }
        }

        return longest == nums.length ? longest - 1 : longest;
    }
}
